cirq-core/cirq/ops/control_values_test.py

```diff
@@ -0,0 +1,117 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pytest
+
+from cirq.ops import control_values as cv
+
+
+def test_init_control_values():
+    tests = [
+        ([], [()]),
+        ([1], [(1,)]),
+        ([[0, 1], 1], [(0, 1), (1, 1)]),
+        ([[[0, 1], [1, 0]]], [(0, 1), (1, 0)]),
+    ]
+    for control_values, want in tests:
+        control_vals = cv.ControlValues(control_values)
+        got = [cv.flatten(product) for product in control_vals]
+        assert want == sorted(got)
+
+
+def test_copy_constructor():
+    tests = [
+        ([], [()]),
+        ([1], [(1,)]),
+        ([[0, 1], 1], [(0, 1), (1, 1)]),
+        ([[[0, 1], [1, 0]]], [(0, 1), (1, 0)]),
+    ]
+    for control_values, want in tests:
+        control_vals = cv.ControlValues([cv.ControlValues(control_values)])
+        got = [cv.flatten(product) for product in control_vals]
+        assert want == sorted(got)
+
+    for control_values, want in tests:
+        values = [cv.ControlValues([cv.ControlValues([val])]) for val in control_values]
+        control_vals = cv.ControlValues(values)
+        got = [cv.flatten(product) for product in control_vals]
+        assert want == sorted(got)
+
+
+def test_constrained_init():
+    tests = [
+        ([[0, 1], [1, 0]], [(0, 1), (1, 0)]),
+        ([[0, 0], [0, 1]], [(0, 0), (0, 1)]),
+        ([[1, 0], [1, 1]], [(0, 1), (1, 1)]),
+    ]
+    for control_values, want in tests:
+        control_vals = cv.ConstrainedVars(control_values)
+        got = [cv.flatten(product) for product in control_vals]
+        assert want == sorted(got)
+
+
+def test_and():
+    originals = [
+        ([], [()]),
+        ([1], [(1,)]),
+        ([[0, 1], 1], [(0, 1), (1, 1)]),
+        ([[[0, 1], [1, 0]]], [(0, 1), (1, 0)]),
+    ]
+    for control_values1, products1 in originals:
+        for control_values2, products2 in originals:
+            control_vals1 = cv.ControlValues(control_values1)
+            control_vals2 = cv.ControlValues(control_values2)
+            want = sorted([v1 + v2 for v1 in products1 for v2 in products2])
+            control_vals1.And(control_vals2)
+            got = sorted([cv.flatten(product) for product in control_vals1])
+            assert want == got
+
+
+def test_slicing_not_supported():
+    control_vals = cv.ControlValues([[[0, 1], [1, 0]]])
+    with pytest.raises(ValueError):
+        control_vals[0:1]  # pylint: disable=pointless-statement
```

Comment:
you can get rid of the pylint warning by doing
`_ = control_vals[0:1]`

Reply:
done